Add setter methods map to WMWorkload and call it for all reqArg parameters

todor-ivanov opened the pull request:
Fixes #12038

#### Status
Ready

#### Description
With the current PR we add the new functionality to call all needed WMWorkload  setter methods based on a full `reqArgs` dictionary passed from the upper level caller and try to set all of parameters at once, rather than calling every single setter method one by one. This is achieved by creating a proper map between methods and possible request arguments. And later validating if the set of arguments passed with `reqArgs` dictionary would properly match the signature of the setter method which is to be called. In the current implementation only a small set of methods is mapped to request parameters :
* setSiteWhiteList
* setSiteBlacklist
* setPriority
(Mostly single argument parametrized methods, but that's  ok for the time being, because those cover perfectly the functionality we need to plug in here)

With this change a path for updating all possible workqueue elements in a single go was opened. In the `WorkQueue` service an additional method was  developed for fetching all possible workqueue elements and update them all with the full set of arguments provided through the `reqArgs` dictionary with the cost of a single database action per WorkQuee element, rather than 3 (or more) separate calls to the database for updating every single element parameter separately. Upon updating all workqueue elements with the new parameters  the WMSpec copy of the given workflow at the workqueue is also updated in a single push using the same logic from above.   

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

todor-ivanov committed:
Add setter methods map to WMWorkload and call it for all reqArg parameters

Add a method in the WorkQueue service for updating all wf elements with all parameters in a sinigle go

Pylint fixes

Remove WQE status filter

Set a WQE status filter as: 'Available', 'Negotiating', 'Acquired'

Handling assignment-approved arguments differently to avoid code duplication

Restore unhandled arguments mechanism

Add back lost else condition to the workload update exception handling

Fix undefined var in error message.

Forgotten raise from None

Call reqMgr api only once && Preserve RequestStatus in the workload.

Add arguments validation for partial workflow parameters update

Use imported ALLOWED_STAT_KEYS during validation

Validate full set of arguments instead of only reqargsDiff

Skip validating update stat arguments

Update workload args only once

Fix bad updateElementsByWorkflow argument

Be explicit in the check for stat update arguments during validation

Review Comments

Pylint fixes

todor-ivanov commented:
@amaltaro  Please take a look at this PR, I think this one  fully covers all the requirements and addresses our fears for affecting  scalability due to increased database calls.

amaltaro commented on src/python/WMCore/WMSpec/WMWorkload.py:
```diff
@@ -68,6 +73,52 @@ class WMWorkloadHelper(PersistencyHelper):
+        #       call the proper setter methods bellow.
+
+        # populate the current instance settersMap
+        self.settersMap['RequestPriority'] = setterTuple('RequestPriority', self.setPriority, inspect.signature(self.setPriority))
```
Given that you are already checking the parameter name and associating it to the actual function (plus inspecting its signature), isn't it more clear to simply call the actual method according to the parameter?

amaltaro replied:
Yes, I understand it. What I am saying is that creating a map and passing a pointer to a function does not look as friendly as I classic if-elif dealing with the 3 parameters.

todor-ivanov replied:
creating a map will gives the option of following the same pattern for adding as many setter calls (wfo parameter setter) as one may wish, by providing a generic method of validating and assembling the calls and not changing the logic for the rest of the function (i.e. not forking with hundreds of elifs, and getting into the trap of code growth by number of custom usecases). Here, the addition of a new setter is just a matter of adding yet another line in the map, following the same pattern as the rest in the map i.e.:
```
a named tuple of: ('reqArg', 'setterFunc', 'setterSignature') 
```
So pretty universal way I'd say. And perfectly serves the exact purpose of what is needed here.

amaltaro replied:
I understand the rationale for implementing something generic, but right now I am not aware of any other spec parameters that will have to be changed, so it's definitely not hundreds of future elifs.

In addition, I am very much adept of the KISS principle, whenever possible and fit. To me, this overly complex code (personally I don't know any of those inspect methods) can easily lead to a bug, or a buggy development.

todor-ivanov replied:
This is exactly to prevent a bug ... not to cause it.  The inspect method is used only to fetch the signature of the underlying function/method  which is to be used for the setter) and upon that simply validate if the number of arguments passed during the call to the function matches the signature and if not it does not trigger the execution, but signals to the caller accordingly.. That's all. 

The difference between this and an `elif` construct is that in the `elif`  you will have to repeat the same lines of call sequence in every section to the call the exact method based on the request parametr, you will have to hardcode the call without validating during the call if the parameters passed match the current signature and if tomorrow someone changes the signature of the setter method but forgets to change the way how we call it here... voila there is your bug, not the other way around.

amaltaro replied:
Well, I actually don't think we need to pre-validate the number of arguments we are passing to a method, because this is already implicit and part of the method execution.

We have a similar method in this very same module that does all of this (used for workflow assignment) and it does not pre-validate number or arguments and/or encapsulate things in try/except, see:
https://github.com/dmwm/WMCore/blob/1f78a64007e58f6c18aff2f0eeee0517d8d99927/src/python/WMCore/WMSpec/WMWorkload.py#L1999

I still thing we are overdoing things with this implementation. I could be though that I am not well educated or evolving with language, so I would appreciate another opinion on this PR (by no means I intend to push responsibility to others!). @vkuznet @khurtado

amaltaro replied:
I guess Andrea wanted to give this answer https://github.com/dmwm/WMCore/pull/12099/files#r1799217523 to this thread instead.

I understand that there are many ways to implement this, either with more or less intellectual knowledge.
So let me ask you this explicitly, @anpicci , do you really think the current implementation is more clear, easier and more maintainable than something like (I don't even think we need try/except here...):
```
        try:
            if 'RequestPriority' in reqArgs:
                self.setPriority(reqArgs['RequestPriority'])
            if 'SiteBlacklist' in reqArgs:
                self.setSiteBlacklist(reqArgs['SiteBlacklist'])
            if 'SiteWhitelist' in reqArgs:
                self.setSiteWhitelist(reqArgs['SiteWhitelist'])
        except Exception as ex:
            msg = f"Failure to update workload parameters. Details: {str(ex)}"
            raise WMWorkloadException(msg) from None
```
?

vkuznet replied:
@amaltaro , @anpicci , @todor-ivanov 

Given that Python (or any programming language) offers multiple ways to achieve the same outcome, evaluating the best approach can often be subjective, depending on factors such as experience, proficiency, or personal style. Consequently, these discussions can become matters of personal preference.

For example, choosing between a full for-loop or list comprehension is a matter of choice. While we can debate which one might be "better" based on various criteria, both serve the same functional purpose. In this instance, I feel the this particular conversation is similar, and rather than debating stylistic choices, the focus should be on whether the code achieves its intended purpose. Otherwise, we risk undermining the role of both developer and reviewer by favoring one approach over another, which I find counterproductive.

If there are concerns about correctness or functionality, I would recommend incorporating unit tests for validation instead of continuing to focus on the specific implementation style.

amaltaro replied:
This is not coding style and personal preferences, but I see it as code readability and sustainability.

Anyhow, thank you for sharing your perspective on this matter. I won't bug people anymore on this.

todor-ivanov replied:
Alan, 

This: 
```
        try:
            if 'RequestPriority' in reqArgs:
                self.setPriority(reqArgs['RequestPriority'])
            if 'SiteBlacklist' in reqArgs:
                self.setSiteBlacklist(reqArgs['SiteBlacklist'])
            if 'SiteWhitelist' in reqArgs:
                self.setSiteWhitelist(reqArgs['SiteWhitelist'])
        except Exception as ex:
            msg = f"Failure to update workload parameters. Details: {str(ex)}"
            raise WMWorkloadException(msg) from None
```
is simply  not equal to this:
```
        for reqArg, argValue in reqArgs.items():
            if not self.settersMap.get(reqArg, None):
                msg = f"Unsupported or missing setter method for updating reqArg: {reqArg}."
                raise WMWorkloadException(msg)
            try:
                self.settersMap[reqArg].setterSignature.bind(argValue)
            except TypeError as ex:
                msg = f"Setter's method signature does not match the method calls we currently support: Error: req{str(ex)}"
                raise WMWorkloadException(msg) from None

        # Now go through the reqArg again and call every setter method according to the map
        for reqArg, argValue in reqArgs.items():
            try:
                self.settersMap[reqArg].setterFunc(argValue)
            except Exception as ex:
                currFrame = inspect.currentframe()
                argsInfo = inspect.getargvalues(currFrame)
                argVals = {arg: argsInfo.locals.get(arg) for arg in argsInfo.args}
                msg = f"Failure while calling setter method {self.settersMap[reqArg].setterFunc.__name__} "
                msg += f"With arguments: {argVals}"
                msg += f"Full exception string: {str(ex)}"
                raise WMWorkloadException(msg) from None
```

At the end, in those both implementations, if the call to the setter method is wrong they will both fail the call, but the later is much more descriptive. It recognizes (at least  3) different possible conditions of failure and forwards the proper message for later debugging, while the former just fails the call and masks all the relevant information for the one who is to chase eventual problems. And that  was the exact reason why I went for this implementation. I know you'd always prefer the shorter version, but sometimes it hides/masks valuable information.

todor-ivanov replied:
Something more, since  this is supposed to be a method triggered by a user's call rather than an operation triggered by the internals of our system, the later method inspects the stack and gives you the values of all parameters in the call. Meaning it shows you the exact (eventually) user mistake.

amaltaro replied:
Talking about inspection and validation. Validation of the user parameters is not supposed to happen at this layer (at the setter call). Validation should be performed in the Service/Request class (which is likely calling the Validation utils).

With that said, I just noticed that we are not validating the site lists, as we usually do for workflow assignment.
So we should add that validation in here as well to protect the system and us from unneded debugging. For reference, this is how we define the site lists and the validation function:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py#L1147

todor-ivanov commented on src/python/WMCore/ReqMgr/Service/Request.py:
```diff
@@ -423,31 +423,19 @@ def _handleNoStatusUpdate(self, workload, request_args, dn):
         workload.saveCouchUrl(workload.specUrl())
 
+        # Commit all Global WorkQueue changes per workflow in a single go:
+        self.gq_service.updateElementsByWorkflow(workload.name(), reqArgs, status=['Available'])
```
> With this call, you will actually change the behavior of priority to get only applied to Available elements.

Well, does it even matter? As we  discussed with you few  days ago - do we actually know if we continue to update any Worqueue Element's parameters at the agent once it is acquired from the GlobalWorkQueue ? If we do, then it matters also for Sitelists as well. So IIRC, during our discussion we decided to stay safe and update only available WQE and not to risk getting into some race conditions related to input data location  etc. If we do not care I can remove the requirement for `status=['Availiable']`. But if we care to propagate the `Priority` information down to the agent for every WQE status - `Aqcuired et. all.`, but we want to update site information only at Global WQ, then we simply  cannot have a mechanism to do it in a single push. The two calls should be separated. So we cannot have them both:
* a single DB action for updating WQE
* different WQE status requirement based on request parameters
(an escape of this is to hardcode this in the Workqueue module logic of `updateelementsByWorkflow`, but you know my opinion on hardcoding requirements in the code.

> In addition, do you ensure now that reqArgs will always be only the supported parameters?

As long as we follow the same logic as we did while validating the allowed parameters in the first set of actions in these calls (meaning to call `validate_request_update_args` before calling the `put` method and hence `_handleNoStatusUpdate`) then we are keeping the same logic as before - hence following the map of allowed actions  from https://github.com/dmwm/WMCore/blob/7ddd036d68d43f3018a6215ff63381fe098fd1fe/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L118-L147. And for this logic we needed mostly those 3 setters methods which I've already put in the `reqArgs` to `setterMethod` map at `updateWorkloadArgs` :
```
self.settersMap['RequestPriority'] = setterTuple('RequestPriority', self.setPriority, inspect.signature(self.setPriority))
self.settersMap['SiteBlacklist'] = setterTuple('SiteBlacklist', self.setSiteBlacklist, inspect.signature(self.setSiteBlacklist))
self.settersMap['SiteWhitelist'] = setterTuple('SiteWhitelist', self.setSiteWhitelist, inspect.signature(self.setSiteWhitelist))
```   

If we  want to be extra detailed and fully exhaustive on any possible `reqarg` per `status` setters, we may also include into this map the eventual methods for all allowed actions per `assignment-approved` status.:
```
    "assignment-approved": ["RequestPriority", "Team", "SiteWhitelist", "SiteBlacklist",
                            "AcquisitionEra", "ProcessingString", "ProcessingVersion",
                            "Dashboard", "MergedLFNBase", "TrustSitelists",
                            "UnmergedLFNBase", "MinMergeSize", "MaxMergeSize",
                            "MaxMergeEvents", "BlockCloseMaxWaitTime",
                            "BlockCloseMaxFiles", "BlockCloseMaxEvents", "BlockCloseMaxSize",
                            "SoftTimeout", "GracePeriod",
                            "TrustPUSitelists", "CustodialSites",
                            "NonCustodialSites", "Override",
                            "SubscriptionPriority"],
 ```   

But for some of them the logic may need to change because some of those setters are parametrized by more than a single argument. While at the same time we already know that for `assignment-approved` we would never call `_handleNoStatusUpdate` ... So that's why, to me it seems safe to proceed only with those 3 setters mapped. Of course,  if you ask me - I am completely up for moving the whole logic to be implemented here in a more generic way .... for all status updates, then get rid of a big chunk of code covering custom cases ... and only make the proper calls to this generic method here from upstream modules (e.g. Request in the current case) But I do not think we will have the time during this line of development  here to do this.

amaltaro replied:
We might be able to distribute this update, but before answering how we can proceed, can you please check the JobUpdater code to check:
1) in which databases are the WQEs updated?
2) does it use any status filter? If so, which statuses?

todor-ivanov replied:
hi @amaltaro, 
ok, Here it is:
* `JobUpdater` at the agent  uses the very same method `updatePriority` https://github.com/dmwm/WMCore/blob/4e0759d5648dfb66de7d7bc135e2975f88fecfbf/src/python/WMCore/Services/WorkQueue/WorkQueue.py#L240 

from the `WMCore.Services.WorkQueue` to update the workflow priority (just as before), here is the actual call:
https://github.com/dmwm/WMCore/blob/4e0759d5648dfb66de7d7bc135e2975f88fecfbf/src/python/WMComponent/JobUpdater/JobUpdaterPoller.py#L151

* It checks for any diff between the priority in WMBS and couch database before taking actions:
https://github.com/dmwm/WMCore/blob/4e0759d5648dfb66de7d7bc135e2975f88fecfbf/src/python/WMComponent/JobUpdater/JobUpdaterPoller.py#L149
and:
https://github.com/dmwm/WMCore/blob/4e0759d5648dfb66de7d7bc135e2975f88fecfbf/src/python/WMComponent/JobUpdater/JobUpdaterPoller.py#L166

* The couchd database to be used is taken from the agent configuration for the component (actually the regular local workqueue component, not the `JobUpdater` config section): 
https://github.com/dmwm/WMCore/blob/4e0759d5648dfb66de7d7bc135e2975f88fecfbf/src/python/WMComponent/JobUpdater/JobUpdaterPoller.py#L50-L51  

which in this case is (taken from one agent configuration):
```
config.WorkQueueManager.dbname = 'workqueue'
```

So the inbox db is indeed configured at the `WorkqueueManager` component at  the agent, but  as  long as I do not see any actual actions  during this update from the `updatepriority` method, I do not think we even touch this DB while updating the WQE priority at the agents.

todor-ivanov commented:
hi @amaltaro I have addressed your comments - Mostly, I removed  the WQE status filter, based on the investigation I did on the `JobUpdater` component.  You may take another look.

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -256,6 +256,41 @@ def updatePriority(self, wf, priority):
+
+        # Update all WorkQueue elements with the parameters provided in a single push
+        if elementsToUpdate:
+            self.updateElements(*elementsToUpdate, **updateParams)
```
I believe I've seen you answering my question about these `updateParams`.
Are we completely 100% sure that the only parameters that will reach this update are one and/or a combination of:
* RequestPriority
* SiteWhitelist
* SiteBlacklist
?
Of course, also considering that we only want to update these values when there is an actual update to the value.

In addition, I would like you to test this in WMAgent with a crafted workflow/scenario, please.

todor-ivanov replied:
> I believe I've seen you answering my question about these updateParams

Yes the answer is the second one in line, here from this comment: https://github.com/dmwm/WMCore/pull/12099#discussion_r1766476657  

> In addition, I would like you to test this in WMAgent with a crafted workflow/scenario, please.

I had  it tested. With a full validation campaign, not a single workflow, in my dev cluster.

amaltaro replied:
I could not find an exact answer to the question above.
Additionally, the code you are deleting had an else statement populating a list of not handled arguments:
https://github.com/dmwm/WMCore/pull/12099/files#diff-120ee6838284a3d1c1799f511da7f147179d0a955f87d0da6fc8b58a8b66c794L440
which makes me believe that it can receive parameters other than only those 3.

About the validation, to properly validate these changes, we need to actually trigger specific scenarios and actions. The standard "campaign-style" validation is not going to expose any issues with this code.

anpicci replied:
@todor-ivanov @amaltaro I don't know if you have already converged on a common ground, but here is my opinion about map vs if-elif-else approach.

Considering that I am not as experienced as both of you with the WMCore software, to me the map approach looks clearer and more understandable than the if-elif-else approach. Indeed, I can read which are the parameters of interest at the very beginning, and from [these lines](https://github.com/dmwm/WMCore/blob/1f78a64007e58f6c18aff2f0eeee0517d8d99927/src/python/WMCore/WMSpec/WMWorkload.py#L100-L108) I can see that only one parameter among RequestPriority, SiteWhitelist, SiteBlacklist can be modified. The all process doesn't require more intellectual effort than the previous if-elif-else approach, and in addition to me sounds more robust and easier to debug (where with "debug" I am also referring to "actually fix a potential issue").

Regarding the possible concerns:
- 

> I believe I've seen you answering my question about these `updateParams`. Are we completely 100% sure that the only parameters that will reach this update are one and/or a combination of:
>     * RequestPriority
>     * SiteWhitelist
>     * SiteBlacklist
>       ?
>       Of course, also considering that we only want to update these values when there is an actual update to the value.
Given the original code, we can assume that this is a complete list of parameters that are supposed to reach this update, provided that what we want to get with this PR doesn't require other parameters to be properly updated





> Additionally, the code you are deleting had an else statement populating a list of not handled arguments:
> https://github.com/dmwm/WMCore/pull/12099/files#diff-120ee6838284a3d1c1799f511da7f147179d0a955f87d0da6fc8b58a8b66c794L440
> which makes me believe that it can receive parameters other than only those 3.
This occurrence should be handled by the try-except in [these lines](https://github.com/dmwm/WMCore/blob/1f78a64007e58f6c18aff2f0eeee0517d8d99927/src/python/WMCore/WMSpec/WMWorkload.py#L100-L108), right @todor-ivanov? @amaltaro do you have any concern that this approach could fail to prevent updating parameters other than  RequestPriority, SiteWhitelist, and SiteBlacklist?

amaltaro replied:
@anpicci I think you meant to send this reply in this thread instead https://github.com/dmwm/WMCore/pull/12099/files#r1766081573 ?

todor-ivanov replied:
To answer again the original question here:

With the latest changes in this commit: https://github.com/dmwm/WMCore/pull/12099/commits/c17e44903d7b515d035928cf2cedb2acde612b33 it is now 100% sure that  if anything but a supported argument update reaches that  point, we will raise the proper  exception and it will be handled accordingly in the caller method

amaltaro commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -256,6 +256,41 @@ def updatePriority(self, wf, priority):
+                                 'reduce': False})
+
+        # Fetch only a list of WorkQueue element Ids && Filter them by allowed status
+        if status:
```
I like that we are not breaking the current behavior of RequestPriority update.

However, thinking twice about this, if we want to make these updates more efficient, we could update it only for elements in Available, Negotiating and Acquired, from the list of potential statuses for WQE:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/DataStructs/WorkQueueElement.py#L14

If you change it, could you please also update the Issue description (because it says only Available).

todor-ivanov replied:
done

amaltaro reviewed (changes requested):
Todor, please find more comments along the code.

Once we are about to converge on this development, I think it is important to test it in a Dev environment (both for central services and WMAgent).

anpicci commented:
@todor-ivanov apart from my comment to @amaltaro 's review, I only suggest to resolve the conflicts on `src/python/WMCore/ReqMgr/Service/Request.py`, if you have not considered yet to address that

anpicci commented:
@vkuznet thank you for the feedback.
@todor-ivanov @amaltaro , I propose to proceeds as follows:
- let's stick with the original Todor's proposal to use maps;
- @amaltaro will focus his review on checking that such implementation is valid and it is consistent with the rest of the system;
- at the same time, @todor-ivanov provides documentation both for the external packages introduced in the PR, and in terms of docstring and in-line comments, when necessary according to both developer and reviewer point of view, such that maintenance of the code is enhanced;
- once Jenkins tests are passed and the PR merged, @todor-ivanov is responsible for fixing issues arising during future release validations that are ascribible to this PR, or, at least, investigate and provide guidances for resolving such issues.

I want to clarify that I would like to adopt these accountability principles to every issue, meaning that this is not a special treatment outlined only for this PR. I will follow up on this during the next group meeting.

Thanks!

amaltaro commented:
It looks like I forgot to update this PR with some previous reviews that have been done here and there, so here we go.

There are some tasks pending from the other related PR, which have been described/discussed in this thread:
https://github.com/dmwm/WMCore/pull/12120#pullrequestreview-2452006874

todor-ivanov commented:
Hi @amaltaro @vkuznet @anpicci 
This  Review  call is mostly for @amaltaro and also to update the others on the latest development triggered by his review comment here: https://github.com/dmwm/WMCore/pull/12099#issuecomment-2515584579 linking to this one: https://github.com/dmwm/WMCore/pull/12120#pullrequestreview-2452006874 , which we agreed to address in the current issue, rather than prolonging the transient development in the previous one. 

So here is  a quick summary: 

* With this commit: [Restore unhandled arguments mechanism](https://github.com/dmwm/WMCore/pull/12099/commits/8f4ae49a573e8232a0f6899c6ff1cae9f522fbb5) 
I am restoring the previous behavior of stopping any workflow changes in the case of not supported parameters update calls, but still signaling which are those which we do not handle.  
**NOTE:** any non-status update parameters for changes of workflows in  `assignment_approved` status are still treated separately

* With this commit: [Call reqMgr api only once && Preserve RequestStatus in the workload.](https://github.com/dmwm/WMCore/pull/12099/commits/4a8ca0e83b7d9bc4b3bf82939856b7e6efa2c8f0)
I am addressing @amaltaro's request to reduce the calls to Reqmgr APIs to only one and only on demand and as early as possible - during the validation step only. Upon which the workflow status is preserved in the `.request`  of the `workload` object and accessible  through the relevant methods through out the whole rest of the process

*  With this commit: [Add arguments validation for partial workflow parameters update](https://github.com/dmwm/WMCore/pull/12099/commits/e6b0d52b814983dbc16bb509fb6d74a6aa2512b8)
I am addressing the request to validate all the values for those partial workflow parameters Updates

todor-ivanov commented on src/python/WMCore/ReqMgr/Service/Request.py:
```diff
@@ -423,37 +426,25 @@ def _handleNoStatusUpdate(self, workload, request_args, dn):
         workload.saveCouchUrl(workload.specUrl())
 
+        # Commit all Global WorkQueue changes per workflow in a single go:
+        self.gq_service.updateElementsByWorkflow(workload.name(), reqArgs, status=['Available', 'Negotiating', 'Acquired'])
```
hi @amaltaro  the line that I was talking about during the WMCore meeting is this one. It triggers a second call to `workload.updateWorkloadArgs` internally, similarly to what is done on line: [431](https://github.com/dmwm/WMCore/pull/12099/files#diff-120ee6838284a3d1c1799f511da7f147179d0a955f87d0da6fc8b58a8b66c794R431)  at `WMcore.ReqMgr.Service.Request`  from this same PR. 

The internal call is happening  at line [316](https://github.com/dmwm/WMCore/pull/12099/files#diff-d758570ce7f6baddaeefb0bdb4e2015b06e5b7772bf929248b686adc64286233R316) at `WMCore.Services.WorkQueue` again from the very same PR.

You have asked me to implement all the changes to both `reqMgr` (in this case through the `workload` object) and all `WQE`s  in a single push. It is now  possible. All that needs to happen, is to substitute the first from above with the second call from the current line. I have the code change prepared. Just let me know  if I should proceed with it or we should play it safe and we still do the action in two steps.

vkuznet reviewed (approved):
I’ve reviewed the code provided in this PR and, from a coding perspective, I don’t have any specific suggestions for improvement. I noticed the ongoing discussion between Alan and Todor on various related topics; however, I believe my input may not add significant value to that part of the conversation.  

From the standpoint of pure code review, the implementation looks good and is ready to be merged. I don’t have a strong preference between Todor's map/namedtuple/setter approach or Alan's if/else flow, as I view this as a matter of personal or team preference.

todor-ivanov commented:
hi @anpicci 

> @todor-ivanov may you ping us when you will provide the fix to the latest issue?

Yes, I will. Most probably after the training session today.

todor-ivanov commented:
hi @anpicci  @amaltaro here is the solution promised:  https://github.com/dmwm/WMCore/pull/12099/commits/fcc7d6c3525745f06038f2f0b81e543f48f0dcb4

todor-ivanov commented:
@amaltaro  I do not know  if you have started looking into this PR, but I need your opinion at least on:  https://github.com/dmwm/WMCore/pull/12099#discussion_r1876282940

So far the code works just as expected with the fix for stat update  arguments. I've tested  it. On top of it here is the state of one such WQE: 
 * Before Sitelists update:
 ```
 {"_id":"2c07fe74e23bcfb33a6fe27115db496b","_rev":"3-bc9c8a587f56eecaefe67dfa12cc9375","timestamp":1733826411.0451522,"updatetime":1733826766.7949412,"thunker_encoded_json":true,"type":"WMCore.WorkQueue.DataStructs.WorkQueueElement.WorkQueueElement","WMCore.WorkQueue.DataStructs.WorkQueueElement.WorkQueueElement":{"Inputs":{"/JetHT/Run2012C-v1/RAW#e759da30-f693-11e1-847a-842b2b4671d8":["T2_CH_CERN","T2_CH_CERN_P5","T2_CH_CERN_HLT"]},"ParentFlag":false,"ParentData":{},"NumberOfLumis":51,"NumberOfFiles":2,"NumberOfEvents":6320,"Jobs":13,"OpenForNewData":false,"NoInputUpdate":false,"NoPileupUpdate":false,"Status":"Acquired","RequestName":"tivanov_TaskChain_LumiMask_multiRun_SiteListsTest_v7_241210_093515_7169","TaskName":"HLTD","Dbs":"https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader","SiteWhitelist":["T1_US_FNAL","T2_CH_CERN"],"SiteBlacklist":[],"StartPolicy":"Block","EndPolicy":{"policyName":"SingleShot"},"Priority":200000,"PileupData":{},"ProcessedInputs":[],"RejectedInputs":[],"ParentQueueId":"tivanov_TaskChain_LumiMask_multiRun_SiteListsTest_v7_241210_093515_7169","SubscriptionId":null,"EventsWritten":0,"FilesProcessed":0,"PercentComplete":0,"PercentSuccess":0,"TeamName":"testbed-vocms0290","ACDC":{},"ChildQueueUrl":"http://vocms0290.cern.ch:5984","ParentQueueUrl":"https://cmsweb-test1.cern.ch/couchdb/workqueue","WMBSUrl":null,"NumOfFilesAdded":0,"Mask":null,"TimestampFoundNewData":1733826410,"CreationTime":1733826411.0451522}}
 ``` 
 * Upon Sitelists updates:
 
 ```
{"_id":"2c07fe74e23bcfb33a6fe27115db496b","_rev":"4-c121956f972ed71d0435d05680ece8bb","timestamp":1733826411.0451522,"updatetime":1733826766.7949412,"thunker_encoded_json":true,"type":"WMCore.WorkQueue.DataStructs.WorkQueueElement.WorkQueueElement","WMCore.WorkQueue.DataStructs.WorkQueueElement.WorkQueueElement":{"Inputs":{"/JetHT/Run2012C-v1/RAW#e759da30-f693-11e1-847a-842b2b4671d8":["T2_CH_CERN","T2_CH_CERN_P5","T2_CH_CERN_HLT"]},"ParentFlag":false,"ParentData":{},"NumberOfLumis":51,"NumberOfFiles":2,"NumberOfEvents":6320,"Jobs":13,"OpenForNewData":false,"NoInputUpdate":false,"NoPileupUpdate":false,"Status":"Acquired","RequestName":"tivanov_TaskChain_LumiMask_multiRun_SiteListsTest_v7_241210_093515_7169","TaskName":"HLTD","Dbs":"https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader","SiteWhitelist":["T2_CH_CERN"],"SiteBlacklist":["T1_US_FNAL"],"StartPolicy":"Block","EndPolicy":{"policyName":"SingleShot"},"Priority":200000,"PileupData":{},"ProcessedInputs":[],"RejectedInputs":[],"ParentQueueId":"tivanov_TaskChain_LumiMask_multiRun_SiteListsTest_v7_241210_093515_7169","SubscriptionId":null,"EventsWritten":0,"FilesProcessed":0,"PercentComplete":0,"PercentSuccess":0,"TeamName":"testbed-vocms0290","ACDC":{},"ChildQueueUrl":"http://vocms0290.cern.ch:5984","ParentQueueUrl":"https://cmsweb-test1.cern.ch/couchdb/workqueue","WMBSUrl":null,"NumOfFilesAdded":0,"Mask":null,"TimestampFoundNewData":1733826410,"CreationTime":1733826411.0451522}}
 ```

Shortly speaking both ReqMgr and the WorkQueue elements have been updated after the arguments have been properly validated  not only for format but also for values type and content. the validation process concerns only no status update actions. 


I am still keeping all those commits unsquashed, only because I am waiting for your answer to the above question and to preserve  the different steps of the solution to be visible during the review process. So if you think no further optimisations  are needed I am ready to squash them and get it ready for merge. If you think I should push further for optimizing these calls to make them in one go I am ready do that as well.

todor-ivanov commented:
And to put it: https://github.com/dmwm/WMCore/pull/12099#discussion_r1876282940 in perspective, this is what I am talking about: [Update workload args only once](https://github.com/dmwm/WMCore/pull/12099/commits/93d0c3b0513074795b500711b0414eef599191a1)


And here are the WQE updated:
* Before:
```
{"_id":"d70157fad5e8778da29c600630bf8521","_rev":"3-b9e5e4642aba0e18b0644a626821d2c9","timestamp":1733826378.292129,"updatetime":1733826385.1671724,"thunker_encoded_json":true,"type":"WMCore.WorkQueue.DataStructs.WorkQueueElement.WorkQueueElement","WMCore.WorkQueue.DataStructs.WorkQueueElement.WorkQueueElement":{"Inputs":{"/NoBPTX/Run2018D-12Nov2019_UL2018-v1/MINIAOD#a9c70800-f936-4c6c-a3a8-c6fa4c9421d0":["T2_CH_CERN","T2_CH_CERN_P5","T2_CH_CERN_HLT"]},"ParentFlag":false,"ParentData":{},"NumberOfLumis":15,"NumberOfFiles":1,"NumberOfEvents":2839,"Jobs":1,"OpenForNewData":false,"NoInputUpdate":false,"NoPileupUpdate":false,"Status":"Acquired","RequestName":"tivanov_ReReco_LumiMask_SiteListsTest_v7_241210_093504_2777","TaskName":"DataProcessing","Dbs":"https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader","SiteWhitelist":["T1_US_FNAL","T2_CH_CERN"],"SiteBlacklist":[],"StartPolicy":"Block","EndPolicy":{"policyName":"SingleShot"},"Priority":600000,"PileupData":{},"ProcessedInputs":[],"RejectedInputs":[],"ParentQueueId":"tivanov_ReReco_LumiMask_SiteListsTest_v7_241210_093504_2777","SubscriptionId":null,"EventsWritten":0,"FilesProcessed":0,"PercentComplete":0,"PercentSuccess":0,"TeamName":"testbed-vocms0290","ACDC":{},"ChildQueueUrl":"http://vocms0290.cern.ch:5984","ParentQueueUrl":"https://cmsweb-test1.cern.ch/couchdb/workqueue","WMBSUrl":null,"NumOfFilesAdded":0,"Mask":null,"TimestampFoundNewData":1733826378,"CreationTime":1733826378.292129}}
```
* After:
```
{"_id":"d70157fad5e8778da29c600630bf8521","_rev":"4-49f3b4a5c45a199783d2347b1a608d3a","timestamp":1733826378.292129,"updatetime":1733826385.1671724,"thunker_encoded_json":true,"type":"WMCore.WorkQueue.DataStructs.WorkQueueElement.WorkQueueElement","WMCore.WorkQueue.DataStructs.WorkQueueElement.WorkQueueElement":{"Inputs":{"/NoBPTX/Run2018D-12Nov2019_UL2018-v1/MINIAOD#a9c70800-f936-4c6c-a3a8-c6fa4c9421d0":["T2_CH_CERN","T2_CH_CERN_P5","T2_CH_CERN_HLT"]},"ParentFlag":false,"ParentData":{},"NumberOfLumis":15,"NumberOfFiles":1,"NumberOfEvents":2839,"Jobs":1,"OpenForNewData":false,"NoInputUpdate":false,"NoPileupUpdate":false,"Status":"Acquired","RequestName":"tivanov_ReReco_LumiMask_SiteListsTest_v7_241210_093504_2777","TaskName":"DataProcessing","Dbs":"https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader","SiteWhitelist":["T2_CH_CERN"],"SiteBlacklist":["T1_US_FNAL"],"StartPolicy":"Block","EndPolicy":{"policyName":"SingleShot"},"Priority":600000,"PileupData":{},"ProcessedInputs":[],"RejectedInputs":[],"ParentQueueId":"tivanov_ReReco_LumiMask_SiteListsTest_v7_241210_093504_2777","SubscriptionId":null,"EventsWritten":0,"FilesProcessed":0,"PercentComplete":0,"PercentSuccess":0,"TeamName":"testbed-vocms0290","ACDC":{},"ChildQueueUrl":"http://vocms0290.cern.ch:5984","ParentQueueUrl":"https://cmsweb-test1.cern.ch/couchdb/workqueue","WMBSUrl":null,"NumOfFilesAdded":0,"Mask":null,"TimestampFoundNewData":1733826378,"CreationTime":1733826378.292129}}
```

So it works! 

@amaltaro - your turn

amaltaro commented on src/python/WMCore/WMSpec/WMWorkload.py:
```diff
@@ -1971,6 +2054,26 @@ def validateArgumentForAssignment(self, schema):
         validateArgumentsUpdate(schema, argumentDefinition)
         return
 
+    def validateArgumentsPartialUpdate(self, schema):
```
This is probably a good idea to write tests too as well.

amaltaro reviewed (changes requested):
@todor-ivanov these changes are looking good in general. I left a few comments along the code that would be better to get them addressed though.

todor-ivanov commented on src/python/WMCore/ReqMgr/Service/Request.py:
```diff
@@ -413,6 +414,8 @@ def _handleNoStatusUpdate(self, workload, request_args, dn):
         """
         reqArgs = deepcopy(request_args)
+        reqStatus = workload.getStatus()
+        cherrypy.log(f"CurrentRequest status: {reqStatus}")
```
why?

todor-ivanov commented on src/python/WMCore/ReqMgr/Service/Request.py:
```diff
@@ -423,37 +426,19 @@ def _handleNoStatusUpdate(self, workload, request_args, dn):
-        if reqArgsNothandled:
+        try:
+            # Commit all Global WorkQueue changes per workflow in a single go:
+            self.gq_service.updateElementsByWorkflow(workload, reqArgs, status=['Available', 'Negotiating', 'Acquired'])
```
The `WMWorkloadUnhandledException` is telling you if there have been any unhandled  arguments. It is thrown when we call  `workload.updateWorkloadArgs` from within `updateElementsByWorkflow` - here: https://github.com/dmwm/WMCore/pull/12099/files#diff-d758570ce7f6baddaeefb0bdb4e2015b06e5b7772bf929248b686adc64286233R316 

And I did it so because of your request during our discussion of the issue to have the `workflow` and the `workqueue elements` being updated in a single push. 
It was not like that just until last night. And I've asked you at least twice ... trying to make all the needed pointers in my comments here: https://github.com/dmwm/WMCore/pull/12099#discussion_r1876282940 and here:  https://github.com/dmwm/WMCore/pull/12099#issuecomment-2539343491 before I proceed with unifying those two steps with this commit:  https://github.com/dmwm/WMCore/pull/12099/commits/93d0c3b0513074795b500711b0414eef599191a1 (hence why I still keep all the commits not squashed)

todor-ivanov replied:
And I do not see how would we recognize if there have been any workqueue elements to be updated until we  query the the workqueue and see if there are any. It have been sequential step and a mandatory one in this method even before I make it all in one go. We  were still always going through it, as long as we were called.  

Usually the parameter `reqArgsDiff` was the one that would have told us if there was anything to be done at this stage within this method, but I remember in the past you asked me to avoid calling the validation twice at the head of this method here. But regardless... , in all the cases of a call to update a request with no `RequestStatus` update we still always return the `reqArgsDiff` from here: https://github.com/dmwm/WMCore/blob/44d2a8a77a42f885bb99f3c4128b823b4e1929ea/src/python/WMCore/ReqMgr/Utils/Validation.py#L111 instead of the originally passed `request_args` and then any actions should have been halted at the very top of this method here: https://github.com/dmwm/WMCore/blob/44d2a8a77a42f885bb99f3c4128b823b4e1929ea/src/python/WMCore/ReqMgr/Service/Request.py#L417-L419 

So  if there was nothing to be updated in the workflow itself - ergo in its `WQE` elements of the given status, we should never reach that point in the code. But if there have been any change with the workflow arguments instead, we will have to proceed with updating the `WQE` elements as well. 

On top of that we update only the ones in the selected statuses, rather than all of them. I think we are safe in what we are doing here.

amaltaro replied:
Given that these changes are now being applied on top of your previous incomplete changes, it is a bit harder to see how things were implemented.

I don't really understand why we call `_handleAssignmentStateTransition()` from this `_handleNoStatusUpdate()` method, as it says that there is not **supposed** to be any status transition. Can you please explain that?

In addition, I believe that this call `updateElementsByWorkflow()` could become not super cheap (few secs?) if workqueue is loaded with documents. So, IMO we should only call it if really necessary. To put it in a different way, a request getting assigned (hence in `assignment-approved`) does not have workqueue elements.

todor-ivanov replied:
> I don't really understand why we call _handleAssignmentStateTransition() from this _handleNoStatusUpdate() method, as it says that there is not supposed to be any status transition. Can you please explain that? 

Because, if you have called the `_handleNoStatusUpdate`  for a workflow that is in `assignment_approved` then you will get  with the request all those full set of parameters that you get during the `assignment_approved` state transition. Which will blow up, because we decided to still support a limited set of parameters for no status update - only `SiteLists` related and `RequestPriority` (which was yet again a wrong decision). This is the root cause of the BUG you reported in the first implementation an did not accept the code back then (even though I have provided this same solution on the very next day).  So this was the solution - the way out of the situation was - in order to avoid rewriting all which is already there in `_handleAssignmentStateTRansition` and assimilating it in  `workload.updateWorkloadArgs` and since the `_handleAssignmentStateTRansition` method has nothing to do with the state transition action itself.... it was safe enough to call it directly here. If you want, I will take this method and will rewrite it under the context of [`workload.updateWorkloadArgs`](https://github.com/dmwm/WMCore/pull/12099/files#diff-b045f03c5ee4e99d47b54e3e08582c4d1a23764cb0b7ba509734628b19c22d86R86). I do not mind. Actually I think this is the proper way to do it. And I've expressed that in the past.

amaltaro replied:
And what is the relationship of a workflow in `assignment_approved` with `updateElementsByWorkflow()`? None, the workflow has not been assigned yet. Calling workqueue is a waste of resources.

If we had an issue (maybe that was with ACDC), making a call to WorkQueue won't resolve that issue. This implementation is wrong. If you want to keep it, keep it.

todor-ivanov replied:
Alan, make up your mind..... 

you asked me to reduce the calls  to worload  and push it all in one go. I twisted my mind to implement it. Before I did it I asked for your feedback twice.... now you tell me this is wrong.

todor-ivanov replied:
If we want to have it all correct we must stop putting exrta background knowledge in the algorithms we follow ... and tie ourselves in a `figure 8` knot of conditions.

todor-ivanov replied:
In order to avoid confusion and blame .... Please put the correct set of conditions here in simple English under which the work queue update elements must be called.  And I will implement it.

amaltaro replied:
As discussed over Zoom:
* `_handleAssignmentStateTransition` is not performing any status transition;
* and the exception `WMWorkloadUnhandledException` is not meant to be raised by `updateElementsByWorkflow()`, but by its internal call `workload.updateWorkloadArgs(updateParams)`, if needed.

Thank you for explaining this Todor. I have no more concerns on this code.

todor-ivanov commented on src/python/WMCore/Services/WorkQueue/WorkQueue.py:
```diff
@@ -284,6 +284,41 @@ def updatePriority(self, wf, priority):
             wmspec.saveCouch(self.hostWithAuth, self.db.name, dummy_values)
         return
 
+    def updateElementsByWorkflow(self, workload, updateParams, status=None):
```
We quickly exchanged a message or two with @vkuznet  through Mattermost and he kindly agreed on my request to keep the current method from this PR here and him adapting his code. Thank you @vkuznet  for this!

amaltaro replied:
Sounds good. Valentin, when you make that pull request, please refer it to the actual GH issue then, which I think it is https://github.com/dmwm/WMCore/issues/12039

todor-ivanov committed:
Unit tests

Unit tests

Unit tests

anpicci reviewed (approved):
Since these commits were discussed on Zoom, I am fine with approving them. There are a few pylint errors to address though @todor-ivanov

todor-ivanov commented:
working on one more Unit test as well

todor-ivanov commented:
@amaltaro , @anpicci  It is all good now, The unit tests are ready and I limited myself to the few pylint fixes that relate the code changes provided with this PR. 

The breaking unit tests are  unstable or not related to the current code change.

 I am merging it now.

@vkuznet  you may consider rebasing  your code on top of master in order to fetch the newly provided `updateWorkloadArgs` method
